plugins/c8y_firmware_plugin/src/tests.rs

```diff
@@ -0,0 +1,775 @@
+        .messages_published_on(&format!(
+            "tedge/{CHILD_DEVICE_ID}/commands/req/firmware_update"
+        ))
+        .await;
+
+    // Publish a c8y_Firmware operation to the plugin
+    broker
+        .publish(
+            "c8y/s/ds",
+            format!(
+                "515,{CHILD_DEVICE_ID},{FIRMWARE_NAME},{FIRMWARE_VERSION},{cloud_firmware_url}"
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Check if the received message payload contains some expected fields and value.
+    let received_message = tedge_command_messages
+        .next()
+        .with_timeout(TEST_TIMEOUT_MS)
+        .await?
+        .expect("No message received.");
+    let received_json: serde_json::Value = serde_json::from_str(&received_message)?;
+    let expected_request_payload = json!({
+        "attempt": 1,
+        "name": FIRMWARE_NAME,
+        "version": FIRMWARE_VERSION,
+        "url": format!("{mock_http_server_host}/tedge/file-transfer/{CHILD_DEVICE_ID}/firmware_update/{file_cache_key}")
+    });
+
+    assert_json_include!(actual: received_json, expected: expected_request_payload);
+    Ok(())
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+#[serial_test::serial]
+async fn handle_request_file_download_failed() -> anyhow::Result<()> {
+    let mut tmp_dir = TempTedgeDir::new();
+    create_required_directories(&mut tmp_dir);
+
+    let broker = mqtt_tests::test_mqtt_broker();
+    let mut c8y_http_client = MockC8YHttpProxy::new();
+    // Key: Download will return Err to imitate downloading failure.
+    c8y_http_client
+        .expect_download_file()
+        .with(
+            predicate::always(),
+            predicate::always(),
+            predicate::always(),
+        )
+        .returning(|_, _, _| Err(SMCumulocityMapperError::RequestTimeout));
+
+    start_firmware_manager(
+        &mut tmp_dir,
+        broker.port,
+        c8y_http_client,
+        DEFAULT_REQUEST_TIMEOUT_SEC,
+    )
+    .await?;
+
+    // Mock download endpoint for the plugin to download a firmware from the cloud
+    let mock_http_server_host = mockito::server_url();
+    let cloud_firmware_url = format!("{mock_http_server_host}/some/cloud/url");
+
+    // Subscribe SmartREST endpoint for child device
+    let mut smartrest_messages = broker
+        .messages_published_on(format!("c8y/s/us/{CHILD_DEVICE_ID}").as_str())
+        .await;
+
+    // Publish a c8y_Firmware operation to the plugin
+    broker
+        .publish(
+            "c8y/s/ds",
+            format!(
+                "515,{CHILD_DEVICE_ID},{FIRMWARE_NAME},{FIRMWARE_VERSION},{cloud_firmware_url}"
+            )
+            .as_str(),
+        )
+        .await?;
+
+    // Assert the c8y_Firmware operation status mapping to EXECUTING(501) and FAILED(502)
+    mqtt_tests::assert_received_all_expected(
+        &mut smartrest_messages,
+        TEST_TIMEOUT_MS,
+        &["501,c8y_Firmware", "502,c8y_Firmware"],
```

Comment:
It'd be nice if the failure reason(at least parts of it) in the 502 message could also be asserted to make sure that the code failed with the expected error and not with something else.